python/ray/dag/dag_node.py

```diff
@@ -58,12 +58,30 @@ def __init__(
         self._bound_other_args_to_resolve: Optional[Dict[str, Any]] = (
             other_args_to_resolve or {}
         )
+
+        # The list of nodes that use this DAG node as input.
+        self._downstream_nodes: List[DAGNode] = []
+        # The list of nodes that this DAG node uses as input.
+        scanner = _PyObjScanner()
+        self._upstream_nodes: List[DAGNode] = scanner.find_nodes(
```

Comment:
The API doc for `find_nodes` is a bit vague: Find top-level DAGNodes.

So here it finds top level nodes or direct upstream nodes? What does top-level nodes mean?

Reply:
Yeah, it's pretty confusing. It's meant to find the immediate upstream nodes. I will update the comment.

Reply:
Updated https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea